search.go

```diff
@@ -559,6 +582,111 @@ func (l *Conn) Search(searchRequest *SearchRequest) (*SearchResult, error) {
 	}
 }
 
+// SearchWithChannel performs a search request and returns all search results
+// via the returned channel as soon as they are received. This means you get
+// all results until an error happens (or the search successfully finished),
+// e.g. for size / time limited requests all are recieved via the channel
+// until the limit is reached.
+func (l *Conn) SearchWithChannel(ctx context.Context, searchRequest *SearchRequest) chan *SearchSingleResult {
+	ch := make(chan *SearchSingleResult)
+	go func() {
+		defer close(ch)
+		if l.IsClosing() {
+			return
+		}
+
+		packet := ber.Encode(ber.ClassUniversal, ber.TypeConstructed, ber.TagSequence, nil, "LDAP Request")
+		packet.AppendChild(ber.NewInteger(ber.ClassUniversal, ber.TypePrimitive, ber.TagInteger, l.nextMessageID(), "MessageID"))
+		// encode search request
+		err := searchRequest.appendTo(packet)
+		if err != nil {
+			ch <- &SearchSingleResult{Error: err}
+			return
+		}
+		l.Debug.PrintPacket(packet)
+
+		msgCtx, err := l.sendMessage(packet)
+		if err != nil {
+			ch <- &SearchSingleResult{Error: err}
+			return
+		}
+		defer l.finishMessage(msgCtx)
+
+		foundSearchSingleResultDone := false
+		for !foundSearchSingleResultDone {
+			select {
+			case <-ctx.Done():
+				l.Debug.Printf("%d: %s", msgCtx.id, ctx.Err().Error())
+				return
+			default:
+				l.Debug.Printf("%d: waiting for response", msgCtx.id)
+				packetResponse, ok := <-msgCtx.responses
+				if !ok {
+					err := NewError(ErrorNetwork, errors.New("ldap: response channel closed"))
+					ch <- &SearchSingleResult{Error: err}
```

Comment:
These writes do not account for a closed channel, which would cause a panic. Since this goroutines does not have a `recover` in the deferred function, the program would crash. 

If we stay with this design, I suggest to add a warning to the function's description to not close the channel outside the library.

Reply:
Good review! I changed the returned channel to receive-only. By doing this, the compiler rejects when the caller closes it.

```
$ go run main.go 
# command-line-arguments
./main.go:62:10: invalid operation: cannot close receive-only channel ch (variable of type <-chan *ldap.SearchSingleResult)
```

Reply:
In any case, I learned `recover` is needed to avoid panic from https://github.com/go-ldap/ldap/actions/runs/5166940415/jobs/9307492985.